dask/dataframe/io/parquet/arrow.py

```diff
@@ -625,6 +654,181 @@ def _construct_parts(
     return parts, stats
 
 
+#
+#  PyArrow Dataset API [PyArrow>=1.0.0] - Private Helper Functions
+#
+
+
+def _get_all_partition_keys(ds):
+    pkeys = defaultdict(list)
+    for file_frag in ds.get_fragments():
+        keys = pa_ds._get_partition_keys(file_frag.partition_expression)
+        for k, v in keys.items():
+            if v not in pkeys[k]:
+                pkeys[k].append(v)
+    return pkeys
+
+
+def _collect_pyarrow_dataset_frags(ds, filters):
+    partition_names = []
+    partition_obj = []
+    partition_keys = {}
+    metadata = []
+    ds_filters = None
+    if filters is not None:
+        ds_filters = pq._filters_to_expression(filters)
+    pkeys = None
+    for file_frag in ds.get_fragments(ds_filters):
+        for rg_frag in file_frag.split_by_row_group(ds_filters, schema=ds.schema):
+            metadata.append(rg_frag)
+            keys = pa_ds._get_partition_keys(rg_frag.partition_expression)
+            if keys:
+                if pkeys is None:
+                    # Get all partition keys without filters...
+                    pkeys = _get_all_partition_keys(ds)
```

Comment:
What's the reason this is done inside the loop? (to avoid running it if there are no partition keys, or if the dataset is empty ?)